### PR TITLE
Fix GPU behavior on Weighted Sum

### PR DESCRIPTION
**Pull request by julhien**
- Add device parameter to `WeightedSum` solver and `SamplerSolver` (and its children)
- Move input parameters to the right device in `derive_unprocessed_input`

Tests now pass on GPU

### CODE REVIEW

**Comment by julhien** [comment]
Current dependencies on/for this PR:
* `dev`
  * **PR #15** <a href="https://app.graphite.dev/github/pr/Angryrou/udao/15?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/Angryrou/udao/15?utm_source=stack-comment).

File: udao/optimization/soo/sampler_solver.py [hunk 1]
```diff
@@ -102,14 +118,16 @@ def filter_on_constraints(
 
         available_indices = np.arange(len(next(iter(input_vars.values()))))
         for constraint in problem.constraints:
-            const_values = problem.apply_function(constraint, input_vars)
+            const_values = problem.apply_function(
+                constraint, input_vars, device=self.device
+            )
```
Comment (Angryrou) [inline udao/optimization/soo/sampler_solver.py:126]: add `.detach().cpu()` here?

**Review by Angryrou — APPROVED** [review]
